File: sdk/go/mics-hook-sdk/micshooksdk/server_test.go
package micshooksdk

import (
	"bytes"
	"context"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestHookServer_Auth_VerifiesSign_WhenRequired(t *testing.T) {
	secretProvider := func(tenantID string) (string, bool) {
		if tenantID == "t1" {
			return "secret", true
		}
		return "", false
	}

	srv := NewHookServer(HookServerOptions{
		RequireSign:        true,
		TenantSecretLookup: secretProvider,
		Auth: func(ctx context.Context, req AuthRequest) (AuthResponse, error) {
			return AuthResponse{
				Meta:     req.Meta,
				Ok:       true,
				UserID:   "u1",
				DeviceID: req.DeviceID,
				Config: &TenantRuntimeConfig{
					HookBaseURL:             "http://hook",
					HeartbeatTimeoutSeconds: 30,
					OfflineBufferTTLSeconds: 300,
					TenantSecret:            "secret",
				},
			}, nil
		},
	})

	meta := HookMeta{TenantID: "t1", RequestID: "r1", TimestampMs: 123, Sign: "", TraceID: "tr1"}
	req := AuthRequest{Meta: meta, Token: "valid:u1", DeviceID: "d1"}

	// Sign rule: HMAC( Encode(req with meta.sign cleared) + requestId + timestampLE64 )
	payloadForSign := EncodeAuthRequest(AuthRequest{Meta: HookMeta{TenantID: meta.TenantID, RequestID: meta.RequestID, TimestampMs: meta.TimestampMs, Sign: "", TraceID: meta.TraceID}, Token: req.Token, DeviceID: req.DeviceID})
	sign := ComputeHookSignBase64("secret", payloadForSign, meta.RequestID, meta.TimestampMs)
	req.Meta.Sign = sign

	body := EncodeAuthRequest(req)

	r := httptest.NewRequest(http.MethodPost, "/auth", bytes.NewReader(body))
	r.Header.Set("Content-Type", "application/protobuf")
	w := httptest.NewRecorder()

	srv.ServeHTTP(w, r)
	if w.Code != 200 {
		t.Fatalf("unexpected status: %d body=%q", w.Code, w.Body.String())
	}
	if ct := w.Header().Get("Content-Type"); ct != "application/protobuf" {
		t.Fatalf("unexpected content-type: %q", ct)
	}
	if len(w.Body.Bytes()) == 0 {
		t.Fatalf("empty response body")
	}
}

File: sdk/node/mics-hook-sdk/src/server.test.ts
import { describe, expect, test } from "vitest";
import { createMicsHookServer } from "./server.js";
import { AuthRequestCodec } from "./proto/mics_hook.js";

function makeReq(path: string, body: Uint8Array, headers?: Record<string, string>) {
  const chunks = [Buffer.from(body)];
  const req: any = {
    method: "POST",
    url: path,
    headers: {
      "content-type": "application/protobuf",
      ...(headers ?? {}),
    },
    on(event: string, cb: (chunk?: any) => void) {
      if (event === "data") {
        for (const c of chunks) cb(c);
      }
      if (event === "end") cb();
    },
  };
  return req;
}

function makeRes() {
  const res: any = {
    statusCode: 200,
    headers: {} as Record<string, string>,
    body: Buffer.alloc(0),
    setHeader(name: string, value: string) {
      this.headers[name.toLowerCase()] = value;
    },
    end(payload?: any) {
      if (payload) this.body = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
    },
  };
  return res;
}

describe("server", () => {
  test("auth endpoint decodes protobuf and encodes protobuf response", async () => {
    const handler = createMicsHookServer({
      requireSign: false,
      tenantSecretProvider: (tenantId) => (tenantId === "t1" ? "s" : ""),
      handlers: {
        auth: async (req) => ({
          meta: req.meta,
          ok: true,
          userId: "u1",
          deviceId: req.deviceId,
          config: { hookBaseUrl: "http://hook", heartbeatTimeoutSeconds: 30, offlineBufferTtlSeconds: 300, tenantMaxConnections: 0, userMaxConnections: 0, tenantMaxMessageQps: 0, tenantSecret: "s" },
          reason: "",
        }),
      },
    });

    const payload = AuthRequestCodec.encode({ meta: { tenantId: "t1", requestId: "r1", timestampMs: 1, sign: "", traceId: "tr1" }, token: "valid:u1", deviceId: "d1" }).finish();
    const req = makeReq("/auth", payload);
    const res = makeRes();

    await handler(req, res);
    expect(res.statusCode).toBe(200);
    expect(res.headers["content-type"]).toContain("application/protobuf");
    expect(res.body.length).toBeGreaterThan(0);
  });
});
